### PR TITLE
GG-25543 Added throttling for thread dumps generation on system failure.

### PR DESCRIPTION
**Pull request by antonovsergey93**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java [hunk 1]
```diff
@@ -16,13 +16,17 @@
 
 package org.apache.ignite.internal.processors.failure;
 
+import java.util.EnumMap;
+import java.util.Map;
+
```
[the remark this answers is not in the document]
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java [hunk 2]
```diff
@@ -65,7 +78,22 @@
     public FailureProcessor(GridKernalContext ctx) {
         super(ctx);
 
-        this.ignite = ctx.grid();
+        ignite = ctx.grid();
+
+        if (igniteDumpThreadsOnFailure) {
```
Comment (slukyano) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java]: Any chance we want to also consinder `systemWorkerBlockedTimeout`? checkpoint read lock timeout is calculated like that: take the systemWorkerBlocked if set, else take failure detection timeout.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java [hunk 3]
```diff
@@ -170,6 +198,36 @@ public synchronized boolean process(FailureContext failureCtx, FailureHandler hn
...
+        assert last != null : "Unknown failure type " + type;
+
+        boolean throttle = curr - last < dumpThreadsTrottlingTimeout;
+
+        if (!throttle)
+            threadDumpPerFailureTypeTime.put(type, curr);
+        else {
+            if (log.isInfoEnabled()) {
+                log.info("Thread dump is hidden due to throttling settings. " +
+                        "Set IGNITE_DUMP_THREADS_ON_FAILURE_THROTTLING_TIMEOUT property to 0 to see all thread dumps");
```
Comment (slukyano) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java]: Need '.' at the end :)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+            FailureContext failureCtx =
+                    new FailureContext(SYSTEM_WORKER_BLOCKED, new Throwable("Failure context error"));
+
+            for (int i = 0; i < 3; i++)
+                ignite.context().failure().process(failureCtx);
+
+            assertTrue(testLog.filteredRecords().isEmpty());
+        }
+        finally {
```
Comment (slukyano) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java]: Is this now preferred to using @Before/@After or beforeTest/afterTest methods?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            if (ignite != null)
+                stopAllGrids();
+        }
+    }
+
+    /**
+     * Custom logger for checking if specified message and thread dump appeared at WARN and ERROR logging level.
+     */
+    private static class TestLogger extends ListeningTestLogger {
```
[the remark this answers is not in the document]
Comment (agura) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java]: Thanks for spotting this! Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java (hunk 2 above)
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java]: We don't consider any other timeouts.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java (hunk 3 above)
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/failure/FailureProcessor.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java (hunk 4 above)
Comment (agura) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java]: It seems there is no rule for it at present. It is easier to find method by names than by annotations.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        startGrid(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
```
Comment (antonovsergey93) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java]: You must clean your resorces before invocation super method. Please move `super.afterTest()` to the end of method.

Comment (agura) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/failure/FailureProcessorThreadDumpThrottlingTest.java]: Done.

**Pull request closed without merging**